### PR TITLE
docs: rewrote advanced guides for containerization

### PR DESCRIPTION
**Pull request by aarnphm**
Signed-off-by: Aaron Pham <29749331+aarnphm@users.noreply.github.com>

## Description

- refactor custom dockerfile templates to containerization since we eventually want to
  move to support OCI-compatible runtime for `bentoml containerize` process.

- add `autoescape=True` to environment to avoid XSS scripting

## Checklist:

- [ ] My code follows the bentoml code style, both `make format` and
  `make lint` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: .yamllint.yml [hunk 1]
```diff
@@ -26,5 +26,7 @@ rules:
   indentation:
     level: warning
     indent-sequences: consistent
+    ignore: |
```
Comment (aarnphm) [inline .yamllint.yml:29]: probably just need to ignore this from yamllint.

File: bentoml/_internal/bento/build_config.py [hunk 2]
```diff
@@ -629,7 +629,7 @@ def __attrs_post_init__(self) -> None:
 
         if use_cuda and use_conda:
             raise BentoMLException(
-                "BentoML does not support using both conda dependencies and setting a CUDA version for GPU. If you need both conda and CUDA, use a custom base image or create a docker_template, see https://docs.bentoml.org/en/latest/concepts/bento.html#custom-base-image-advanced"
+                "BentoML does not support using both conda dependencies and setting a CUDA version for GPU. If you need both conda and CUDA, use a custom base image or create a dockerfile_template, see https://docs.bentoml.org/en/latest/concepts/bento.html#custom-base-image-advanced"
```
Comment (aarnphm) [inline bentoml/_internal/bento/build_config.py:632]: we should add `forbid_extra_keys` to cattrs when it is supported.

File: hooks/pre-commit [hunk 3]
```diff
@@ -17,17 +17,17 @@ setup-cfg-fmt --max-py-version 3.10 "$GIT_ROOT/setup.cfg"
...
 done
 
 INFO "Running branch name validation..."
 if [[ ! "$LOCAL_BRANCH" =~ $VALID_BRANCH_REGEX ]]; then
-	WARN "Branch name %s doesn't follow the predefined format...\n" "$LOCAL_BRANCH"
-	WARN "We aren't going to abort the commit, but we recommend that branch name should adhere to the following format:\n\t%s.\n" "$VALID_BRANCH_REGEX"
-	WARN "To rename your branch to follow the format do \`git branch -m 'feature/%s'\`\n" "$LOCAL_BRANCH"
+	WARN "Branch name $LOCAL_BRANCH doesn't follow the predefined format...\n"
+	WARN "We aren't going to abort the commit, but we recommend that branch name should adhere to the following format:\n\t$VALID_BRANCH_REGEX.\n"
+	WARN "To rename your branch to follow the format do \`git branch -m 'feature/$LOCAL_BRANCH'\`\n"
```
Comment (aarnphm) [inline hooks/pre-commit:31]: Fixes some printing issue.

File: hooks/pre-commit [hunk 4]
```diff
@@ -4,7 +4,7 @@ LC_ALL=C
 GIT_ROOT="$(git rev-parse --show-toplevel)"
 
 LOCAL_BRANCH="$(git rev-parse --abbrev-ref HEAD)"
-VALID_BRANCH_REGEX="^(qa|style|chore|refactor|feat|internal|bugfix|improvement|library|prerelease|release|hotfix)\/[a-zA-Z0-9._-]+$"
+VALID_BRANCH_REGEX="^(qa|style|chore|refactor|feat|internal|bugfix|improvement|library|docs|prerelease|release|hotfix)\/[a-zA-Z0-9._-]+$"
```
Comment (aarnphm) [inline hooks/pre-commit:7]: we should also accept `docs: ...`

File: docs/source/guides/index.rst [hunk 5]
```diff
@@ -14,7 +14,7 @@ into this part of the documentation.
 
   batching
   configuration
-  dockerfile_generation
+  containerization
```
Comment (aarnphm) [inline docs/source/guides/index.rst:17]: since we want to eventually support OCI-spec, probably a good idea to move away from just docker :)

File: bentoml/_internal/io_descriptors/multipart.py [hunk 6]
```diff
@@ -106,7 +106,7 @@ def predict(arr, annotations):
                 |                                                                |
                 |   +--------------------------------------------------------+   |
                 |   |                                                        |   |
-                |   |    Multipart(arr=NumpyNdarray(), annotations=JSON()    |   |
+                |   |    Multipart(arr=NumpyNdarray(), annotations=JSON())   |   |
```
Comment (aarnphm) [inline bentoml/_internal/io_descriptors/multipart.py:109]: ran across this. Oops.

**Review by aarnphm — COMMENTED** [review]
I believe I have address all changes needed here.

File: hooks/pre-commit [hunk 7]
```diff
@@ -17,17 +17,17 @@ setup-cfg-fmt --max-py-version 3.10 "$GIT_ROOT/setup.cfg"
 git diff --name-status | grep setup.cfg &>/dev/null && git add setup.cfg
 
 # Run format scripts
-./scripts/tools/formatter.sh || exit 1
+./scripts/tools/formatter.sh && git add . || exit 1
```
Comment (parano) [inline hooks/pre-commit:20]:
```suggestion
./scripts/tools/formatter.sh || exit 1
```

File: .yamllint.yml [hunk 8]
```diff
@@ -26,5 +26,7 @@ rules:
...
+      .github/workflows/codeql-analysis.yml
```
Comment (sauyon) [inline .yamllint.yml:30]: Let's just fix this? What's wrong with it?

File: bentoml/_internal/bento/gen.py [hunk 9]
```diff
@@ -173,6 +173,7 @@ def generate_dockerfile(
         extensions=["jinja2.ext.do", "jinja2.ext.loopcontrols", "jinja2.ext.debug"],
         trim_blocks=True,
         lstrip_blocks=True,
+        autoescape=True,  # https://docs.huihoo.com/google-ctemplate/auto_escape.html
```
Comment (sauyon) [inline bentoml/_internal/bento/gen.py:176]: Hm, seems like this is made for escaping HTML / JS and not dockerfiles?

File: docs/source/concepts/bento.rst [hunk 10]
```diff
@@ -796,76 +796,70 @@ will build a new image on top of the base_image with the following steps:
...
-"""""""""""""""""""""""""""""
+Dockerfile Template (Beta)
+""""""""""""""""""""""""""
 
-The :code:`docker_template` field gives user the full control over how the
-:code:`Dockerfile` was generated in a Bento. Users can use this field to
-customize Bento's Dockerfile instruction set to suits their needs.
+The :code:`dockerfile_template` field gives user the full control over how the
+:code:`Dockerfile` was generated in a Bento. Users can extend the template used for
+generating Dockerfile in a Bento, to suits their needs.
```
Comment (sauyon) [inline docs/source/concepts/bento.rst:804]:
```suggestion
The :code:`dockerfile_template` field gives the user full control over how the
:code:`Dockerfile` is generated for a Bento by extending the template used by
BentoML.
```

File: docs/source/concepts/bento.rst [hunk 11]
```diff
@@ -796,76 +796,70 @@ will build a new image on top of the base_image with the following steps:
...
 
-First, create a :code:`Dockerfile.template` file next to your :code:`bentofile.yaml` build file.
-This template file is a mixed between a :code:`Dockerfile` and a :code:`Jinja` template file:
+First, create a :code:`Dockerfile.template` file next to your :code:`bentofile.yaml`
+build file. This file should follow the
+`Jinja2 <https://jinja.palletsprojects.com/en/3.1.x/>`_ template language, and extend
+BentoML's base template and blocks. The template should render a valid
+`Dockerfile https://docs.docker.com/engine/reference/builder/`_. For example:
```
Comment (sauyon) [inline docs/source/concepts/bento.rst:810]: Should we call it `Dockerfile.j2`?

Seems like it doesn't need to be saved alongside the `bentofile.yaml`?

```suggestion
First, create a `Jinja2 <https://jinja.palletsprojects.com/en/3.1.x/>`_ template for your Dockerfile.
This file can extend BentoML's base template and blocks. For example:
```

File: docs/source/concepts/bento.rst [hunk 12]
```diff
@@ -796,76 +796,70 @@ will build a new image on top of the base_image with the following steps:
...
    RUN echo "We are running this during bentoml containerize!"
    {% endblock %}
 
-.. note::
-
-   The template file can have extension :code:`.jinja`, :code:`.j2`, or any
-   extensions that Jinja2 can understand.
-
-Then add the path to the given template file to the :code:`docker_template` field in the :code: `bentofile.yaml`:
+Then add the path to the given template file to the :code:`dockerfile_template` field in
```
Comment (sauyon) [inline docs/source/concepts/bento.rst:820]:
```suggestion
Then add the path to your template file to the :code:`dockerfile_template` field in
```

File: docs/source/concepts/bento.rst [hunk 13]
```diff
@@ -796,76 +796,70 @@ will build a new image on top of the base_image with the following steps:
...
-        docker_template: "./Dockerfile.template"
-
-.. note:: 
+        dockerfile_template: "./Dockerfile.template"
 
-   Although we support any local file path, it is preferred that users put the
-   template file in the project directory.
+Now run :code:`bentoml build` to build a new Bento, it will contain a Dockerfile
+generated with the custom template. To confirm the generated Dockerfile work as
+expected, run :code:`bentoml containerize <bento>` to build a docker image with it.
```
Comment (sauyon) [inline docs/source/concepts/bento.rst:830]:
```suggestion
Now run :code:`bentoml build` to build a new Bento. It will contain a Dockerfile
generated with the custom template. To confirm the generated Dockerfile works as
expected, run :code:`bentoml containerize <bento>` to build a docker image with it.
```

File: docs/source/concepts/bento.rst [hunk 14]
```diff
@@ -796,76 +796,70 @@ will build a new image on top of the base_image with the following steps:
...
 
-Now to see the result generated Dockerfile, do the following:
+    .. code-block:: bash
 
-.. code-block:: bash
-
-   bentoml build && cat $(bentoml get <bento> -o path)/env/docker/Dockerfile
-
-Run :code:`bentoml containerize <bento>` to confirm the generated Dockerfile work as expected.
+        cat $(bentoml get MY_BENTO_NAME:latest -o path)/env/docker/Dockerfile
```
Comment (sauyon) [inline docs/source/concepts/bento.rst:840]:
```suggestion
        cat "$(bentoml get MY_BENTO_NAME:latest -o path)/env/docker/Dockerfile"
```

File: docs/source/concepts/bento.rst [hunk 15]
```diff
@@ -796,76 +796,70 @@ will build a new image on top of the base_image with the following steps:
...
+expected, run :code:`bentoml containerize <bento>` to build a docker image with it.
 
-   .. code:: yaml
+.. dropdown:: View the generated Dockerfile content
+    :icon: code
 
-      docker:
-          docker_template: "~/workspace/Dockerfile.template"
+    During development and debugging, you may want to see the generated Dockerfile
+    after a Bento is built. Here's shortcut for that:
```
Comment (sauyon) [inline docs/source/concepts/bento.rst:836]:
```suggestion
    During development and debugging, you may want to see the generated Dockerfile.
    Here's shortcut for that:
```

File: docs/source/concepts/bento.rst [hunk 16]
```diff
@@ -796,76 +796,70 @@ will build a new image on top of the base_image with the following steps:
...
-
-   bentoml build && cat $(bentoml get <bento> -o path)/env/docker/Dockerfile
-
-Run :code:`bentoml containerize <bento>` to confirm the generated Dockerfile work as expected.
+        cat $(bentoml get MY_BENTO_NAME:latest -o path)/env/docker/Dockerfile
 
 .. seealso::
 
-    :ref:`Dockerfile Generation <guides/dockerfile_generation:Dockerfile generation>` to learn more about how BentoML generates Dockerfile for Bentos.
+   :ref:`Bento Containerization <guides/containerization:Containerization>` to learn more about how BentoML containerize a Bento.
```
Comment (sauyon) [inline docs/source/concepts/bento.rst:844]:
```suggestion
   :ref:`Bento Containerization <guides/containerization:Containerization>`: more information about how BentoML containerizes a Bento.
```

File: docs/source/concepts/bento.rst [hunk 17]
```diff
@@ -796,76 +796,70 @@ will build a new image on top of the base_image with the following steps:
         bentoml containerize iris_classifier:latest --platform=linux/amd64
 
 
-Docker Template (Danger Zone)
-"""""""""""""""""""""""""""""
+Dockerfile Template (Beta)
```
Comment (sauyon) [inline docs/source/concepts/bento.rst:799]: Should we move this into an advanced guide?

File: hooks/pre-commit (hunk 4 above)
Comment (sauyon) [reply hooks/pre-commit:7]: I don't really think this check is necessary.

**Review by sauyon — COMMENTED** [review]
Haven't reviewed the containerization doc yet, but here's some early stuff.

File: .yamllint.yml (hunk 8 above)
Comment (aarnphm) [reply .yamllint.yml:30]: spacing. i dont think it is relevant

File: bentoml/_internal/bento/gen.py (hunk 9 above)
Comment (aarnphm) [reply bentoml/_internal/bento/gen.py:176]: i guess but bandit is complaining about it.

File: docs/source/concepts/bento.rst (hunk 11 above)
Comment (aarnphm) [reply docs/source/concepts/bento.rst:810]: the reason for L806 is that we want to enforce this practice to our users. Eventhough we accept the any path to templates, we should just mention to them that they should put it in the build context.

File: docs/source/concepts/bento.rst (hunk 17 above)
Comment (aarnphm) [reply docs/source/concepts/bento.rst:799]: no.

File: hooks/pre-commit (hunk 4 above)
Comment (aarnphm) [reply hooks/pre-commit:7]: lets just remove it

File: .github/workflows/codeql-analysis.yml [hunk 18]
```diff
@@ -24,17 +24,17 @@ jobs:
         language: [ 'python' ]
 
     steps:
-    - name: Checkout repository
-      uses: actions/checkout@v2
+      - name: Checkout repository
```
Comment (parano) [inline .github/workflows/codeql-analysis.yml:27]: don't commit unrelated changes

File: bentoml/_internal/bento/gen.py (hunk 9 above)
Comment (aarnphm) [reply bentoml/_internal/bento/gen.py:176]: oh i ran [bandit](https://github.com/PyCQA/bandit) locally. I will remove this for now.

File: docs/source/guides/containerization.rst [hunk 19]
```diff
@@ -0,0 +1,596 @@
...
+
+   {% endblock %}
+
+Then add the following to your :code:`bentofile.yaml`:
+
+.. code-block:: yaml
+
+   docker:
+     system_packages:
+       - nginx
```
Comment (parano) [inline docs/source/guides/containerization.rst:87]:
```suggestion
```

Comment (parano) [reply docs/source/guides/containerization.rst:87]: this will be ignored right?

File: .github/workflows/codeql-analysis.yml (hunk 18 above)
Comment (aarnphm) [reply .github/workflows/codeql-analysis.yml:27]: Maybe we should have a fix for these type of lint error soon. Will remove this for now.

File: docs/source/guides/containerization.rst [hunk 20]
```diff
@@ -0,0 +1,596 @@
...
+
+   You can also provide :code:`--progress plain` to see the progress from
+   `buildkit <https://github.com/moby/buildkit>`_ in plain text
+
+   .. code-block:: yaml
+
+      bentoml containerize --progress plain <bento>:<tag>`
+
+Providing AWS credentials to Bentos
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (parano) [inline docs/source/guides/containerization.rst:108]:
```suggestion
Access AWS credentials during image build
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: docs/source/guides/containerization.rst [hunk 21]
```diff
@@ -0,0 +1,596 @@
...
+3. A big difference with :ref:`base image <concepts/bento:Docker Options Table>` features is that you don't have to setup a custom base image and then push it to a remote registry.
+
+How it works
+------------
+
+To focus on how to create a custom Dockerfile template, the following examples
+are provided:
+
+1. :ref:`guides/containerization:Building binary into Bentos`
+2. :ref:`guides/containerization:Providing AWS credentials to Bentos`
```
Comment (parano) [inline docs/source/guides/containerization.rst:31]:
```suggestion
2. :ref:`guides/containerization:Access AWS credentials during image build`
```

File: docs/source/guides/containerization.rst [hunk 22]
```diff
@@ -0,0 +1,596 @@
...
+       - nginx
+     dockerfile_template: ./Dockerfile.template
+
+Proceed to build your Bento with :code:`bentoml build` and containerize with :code:`bentoml containerize`:
+
+.. code-block:: bash
+
+   bentoml build
+
+   bentoml containerize <bento>:<tag>`
```
Comment (parano) [inline docs/source/guides/containerization.rst:96]:
```suggestion
   bentoml containerize <bento>:<tag>
```

File: docs/source/guides/containerization.rst [hunk 23]
```diff
@@ -0,0 +1,596 @@
+================
+Containerization
+================
+
+This guide describes the containerization process and how users can
+define a custom Dockerfile templates to customize Bento container. 
+
+
+.. warning::
```
Comment (parano) [inline docs/source/guides/containerization.rst:9]:
```suggestion
.. note::
```

File: docs/source/guides/containerization.rst (hunk 19 above)
Comment (aarnphm) [reply docs/source/guides/containerization.rst:87]: actually no

File: docs/source/guides/containerization.rst [hunk 24]
```diff
@@ -0,0 +1,596 @@
...
+are provided:
+
+1. :ref:`guides/containerization:Building binary into Bentos`
+2. :ref:`guides/containerization:Providing AWS credentials to Bentos`
+3. :ref:`guides/containerization:Installing custom CUDA version with conda`
+
+Building binary into Bentos
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Let's start with an example that builds an `Ingress <https://github.com/kubernetes/ingress-nginx>`_ binary into a Bento.
```
[the remark this answers is not in the document]
Comment (aarnphm) [reply docs/source/guides/containerization.rst:37]: sounds good i will add this accordingly.

File: docs/source/guides/containerization.rst [hunk 25]
```diff
@@ -0,0 +1,596 @@
...
+   bentoml containerize <bento>:<tag>`
+
+.. tip:: 
+
+   You can also provide :code:`--progress plain` to see the progress from
+   `buildkit <https://github.com/moby/buildkit>`_ in plain text
+
+   .. code-block:: yaml
+
+      bentoml containerize --progress plain <bento>:<tag>`
```
Comment (aarnphm) [inline docs/source/guides/containerization.rst:105]:
```suggestion
      bentoml containerize --progress plain <bento>:<tag>
```

File: docs/source/guides/containerization.rst (hunk 22 above)
Comment (aarnphm) [inline docs/source/guides/containerization.rst]:
```suggestion
   bentoml containerize <bento>:<tag>
```

**Review by aarnphm — COMMENTED** [review]
Some reason the changes are not reflected.

File: docs/source/concepts/bento.rst (hunk 11 above)
Comment (sauyon) [reply docs/source/concepts/bento.rst:810]: Ah, I see. We should just make it clear it's a recommendation, then; I can see use-cases where people want to have multiple bentos use the same dockerfile template.

File: docs/source/guides/containerization.rst [hunk 26]
```diff
@@ -0,0 +1,602 @@
+================
+Containerization
```
Comment (parano) [inline docs/source/guides/containerization.rst:2]: rename this title to "Custom Dockerfile"?

This guide is not specific to Containerization - it's about advanced customization of the generated dockerfile

File: docs/source/guides/containerization.rst [hunk 27]
```diff
@@ -0,0 +1,602 @@
...
+================
+
+This guide describes the containerization process and how users can
+define a custom Dockerfile templates to customize Bento container.
```
Comment (parano) [inline docs/source/guides/containerization.rst:6]: This guides describes how to customize the Dockerfile generated for containerizing a Bento, by extending BentoML's built-in templates. This is an advanced feature for user to customize container environment that are not directly supported in BentoML. See `Bento Docker Options ` doc to learn about basic options for configuring the Bento generated Docker image.

File: docs/source/guides/containerization.rst [hunk 28]
```diff
@@ -0,0 +1,602 @@
...
+
+.. note::
+
+   BentoML make uses of `Jinja2 <https://jinja.palletsprojects.com/en/3.1.x/>`_ to enable custom Dockerfile template.
+
+   However, This section is not meant to be a complete reference on Jinja2.
+   It is meant to give a quick overview of how Jinja2 is used in conjunction with BentoML.
+   For any reference on Jinja2 please refers to their `Templates Design Documentation <https://jinja.palletsprojects.com/en/3.1.x/templates/>`_.
+
+Why do you need this?
```
Comment (parano) [inline docs/source/guides/containerization.rst:17]:
```suggestion
Why you may need this?
```

File: docs/source/guides/containerization.rst [hunk 29]
```diff
@@ -0,0 +1,602 @@
...
+------------
+
+To focus on how to create a custom Dockerfile template, the following examples
+are provided:
+
+1. :ref:`guides/containerization:Building binary into Bentos`
+2. :ref:`guides/containerization:Access AWS credentials during image build`
+3. :ref:`guides/containerization:Installing custom CUDA version with conda`
+
+Building binary into Bentos
```
Comment (parano) [inline docs/source/guides/containerization.rst:34]: Rename this section to something like "Building Tensorflow custom op"?

File: docs/source/guides/containerization.rst (hunk 26 above)
Comment (aarnphm) [reply docs/source/guides/containerization.rst:2]: sure. I don't have strong opinion on this.

**Review by ssheng — APPROVED** [review]
